titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/DefaultKubeJobManagementReconciler.java

```diff
@@ -0,0 +1,286 @@
+
+    private final DirectProcessor<ContainerEvent> v3ContainerEventProcessor = DirectProcessor.create();
+    private final FluxSink<ContainerEvent> v3ContainerEventSink = v3ContainerEventProcessor.sink(FluxSink.OverflowStrategy.IGNORE);
+
+    private final DirectProcessor<PodEvent> podEventProcessor = DirectProcessor.create();
+    private final FluxSink<PodEvent> podEventSink = podEventProcessor.sink(FluxSink.OverflowStrategy.IGNORE);
+
+    private final Map<OrphanedKind, Gauge> orphanedTaskGauges;
+
+    private ScheduleReference schedulerRef;
+
+    @Inject
+    public DefaultKubeJobManagementReconciler(MesosConfiguration mesosConfiguration,
+                                              KubeApiFacade kubeApiFacade,
+                                              V3JobOperations v3JobOperations,
+                                              TitusRuntime titusRuntime) {
+        this.mesosConfiguration = mesosConfiguration;
+        this.kubeApiFacade = kubeApiFacade;
+        this.v3JobOperations = v3JobOperations;
+        this.clock = titusRuntime.getClock();
+        this.titusRuntime = titusRuntime;
+
+        Registry registry = titusRuntime.getRegistry();
+
+        this.orphanedTaskGauges = Stream.of(OrphanedKind.values()).collect(Collectors.toMap(
+                Function.identity(),
+                kind -> registry.gauge(MetricConstants.METRIC_KUBERNETES + "orphanedTasks", "kind", kind.name())
+        ));
+    }
+
+    @Activator
+    public void enterActiveMode() {
+        ScheduleDescriptor scheduleDescriptor = ScheduleDescriptor.newBuilder()
+                .withName("reconcileNodesAndPods")
+                .withDescription("Reconcile nodes and pods")
+                .withInitialDelay(Duration.ofMillis(mesosConfiguration.getReconcilerInitialDelayMs()))
+                .withInterval(Duration.ofMillis(mesosConfiguration.getReconcilerIntervalMs()))
+                .withTimeout(Duration.ofMinutes(5))
+                .build();
+        this.schedulerRef = titusRuntime.getLocalScheduler().schedule(
+                scheduleDescriptor,
+                e -> reconcile(),
+                ExecutorsExt.namedSingleThreadExecutor(DefaultKubeJobManagementReconciler.class.getSimpleName())
+        );
+    }
+
+    @Deactivator
+    @PreDestroy
+    public void shutdown() {
+        Evaluators.acceptNotNull(schedulerRef, ScheduleReference::cancel);
+    }
+
+    @Override
+    public Flux<ContainerEvent> getV3ContainerEventSource() {
+        return v3ContainerEventProcessor.compose(ReactorExt.badSubscriberHandler(logger));
+    }
+
+    @Override
+    public Flux<PodEvent> getPodEventSource() {
+        return podEventProcessor.compose(ReactorExt.badSubscriberHandler(logger));
+    }
+
+    private void reconcile() {
+        if (!mesosConfiguration.isReconcilerEnabled() || !kubeApiFacade.getNodeInformer().hasSynced() || !kubeApiFacade.getPodInformer().hasSynced()) {
+            return;
```

Comment:
log something and/or emit a metric so we know when reconciliation is not running? Or alternatively, emit a metric when it runs